base_layer/core/src/base_node/states/horizon_state_sync/state_sync.rs

```diff
@@ -0,0 +1,705 @@
+    ) -> u64
+    {
+        let network_tip_height = self.network_metadata.height_of_longest_chain.unwrap_or(0);
+        let horizon_height = local_metadata.horizon_block(network_tip_height);
+        let horizon_sync_height_offset = shared.config.horizon_sync_config.horizon_sync_height_offset;
+        cmp::min(horizon_height + horizon_sync_height_offset, network_tip_height)
+    }
+
+    pub fn network_metadata(&self) -> &ChainMetadata {
+        &self.network_metadata
+    }
+
+    pub fn sync_peers(&self) -> &[NodeId] {
+        &self.sync_peers
+    }
+}
+
+impl PartialEq for HorizonStateSync {
+    fn eq(&self, other: &Self) -> bool {
+        self.sync_peers == other.sync_peers && self.network_metadata == other.network_metadata
+    }
+}
+
+struct HorizonStateSynchronization<'a, 'b, B> {
+    shared: &'a mut BaseNodeStateMachine<B>,
+    local_metadata: ChainMetadata,
+    sync_peers: &'b mut Vec<NodeId>,
+    horizon_sync_height: u64,
+}
+
+impl<B: BlockchainBackend + 'static> HorizonStateSynchronization<'_, '_, B> {
+    pub async fn synchronize(&mut self) -> Result<(), HorizonSyncError> {
+        let local_tip_height = self.local_metadata.height_of_longest_chain.unwrap_or(0);
+        debug!(
+            target: LOG_TARGET,
+            "Syncing from height {} to horizon sync height {}.", local_tip_height, self.horizon_sync_height
+        );
+
+        let tip_header = async_db::fetch_tip_header(self.db()).await?;
+        if tip_header.height >= self.horizon_sync_height {
+            return Ok(());
+        }
+
+        // During horizon state syncing the blockchain backend will be in an inconsistent state until the entire horizon
+        // state has been synced. Reset the local chain metadata will limit other nodes and local service from
+        // requesting data while the horizon sync is in progress.
+        trace!(target: LOG_TARGET, "Resetting chain metadata.");
+        self.reset_chain_metadata_to_genesis().await?;
+        trace!(target: LOG_TARGET, "Synchronizing headers");
+        self.synchronize_headers(tip_header.height).await?;
+        trace!(target: LOG_TARGET, "Synchronizing kernels");
+        self.synchronize_kernels().await?;
+        trace!(target: LOG_TARGET, "Check the deletion state of current UTXOs");
+        self.check_state_of_current_utxos(tip_header.height).await?;
+        trace!(target: LOG_TARGET, "Synchronizing UTXOs and RangeProofs");
+        self.synchronize_utxos_and_rangeproofs().await?;
+        trace!(target: LOG_TARGET, "Finalizing horizon synchronizing");
+        self.finalize_horizon_sync().await?;
+
+        Ok(())
+    }
+
+    async fn synchronize_headers(&mut self, tip_height: u64) -> Result<(), HorizonSyncError> {
+        let config = self.shared.config.horizon_sync_config;
+
+        let block_height_range = ((tip_height + 1)..=self.horizon_sync_height).collect::<Vec<_>>();
+        for block_nums in block_height_range.chunks(config.header_request_size) {
```

Comment:
This code should be quite similar in the block sync state. Maybe it should be a common function that both call?

Reply:
Agreed - I've got some ideas on how to make it better - this code was moved from PR #2004

Reply:
Perhaps in a subsequent PR? Pruned mode is still in active development